File: adapter/js/aggregator.js
'use strict';

/**
 * Aggregator
 *
 * Groups multiple Zoom connection segments (same person, same meeting)
 * into a single attendance record with total minutes per meeting-half.
 *
 * Accepts optional maps to override break points and effective start times.
 */
const Aggregator = {

    /**
     * @param {Array} meetings
     * @param {Map<number, Date>} [breakPoints]
     * @param {Map<number, Date>} [effectiveStarts]
     * @returns {Array}
     */
    aggregate(meetings, breakPoints, effectiveStarts) {
        const records = [];

        for (let i = 0; i < meetings.length; i++) {
            const meeting = meetings[i];
            if (meeting.segments.length === 0) continue;

            const effStart = (effectiveStarts && effectiveStarts.get(i)) || meeting.startTime;

            const bp = breakPoints ? breakPoints.get(i) : null;
            let firstEnd, secondStart;

            if (bp) {
                firstEnd = bp;
                secondStart = bp;
            } else {
                const breakInfo = BreakDetector.detect(meeting);
                if (breakInfo) {
                    firstEnd = breakInfo.breakStart;
                    secondStart = breakInfo.breakEnd;
                } else {
                    const mid = new Date(
                        (effStart.getTime() + meeting.endTime.getTime()) / 2
                    );
                    firstEnd = mid;
                    secondStart = mid;
                }
            }

            const durationFirst = (firstEnd - effStart) / 60000;
            const durationSecond = (meeting.endTime - secondStart) / 60000;

            const groups = this._groupByParticipant(meeting.segments);

            for (const [, segments] of groups) {
                const first = segments[0];

                let minFirst = 0;
                let minSecond = 0;

                for (const seg of segments) {
                    minFirst += this._overlap(seg.joinTime, seg.leaveTime, effStart, firstEnd);
                    minSecond += this._overlap(seg.joinTime, seg.leaveTime, secondStart, meeting.endTime);
                }

                records.push({
                    course: meeting.course,
                    meetingId: meeting.meetingId,
                    date: meeting.startTime,
                    meetingStart: meeting.startTime,
                    meetingEnd: meeting.endTime,
                    meetingDuration: meeting.durationMinutes,
                    firstName: first.firstName,
                    lastName: first.lastName,
                    email: first.email,
                    minutesFirstHalf: this._round1(minFirst),
                    minutesSecondHalf: this._round1(minSecond),
                    durationFirstHalf: this._round1(durationFirst),
                    durationSecondHalf: this._round1(durationSecond),
                    totalMinutes: this._round1(minFirst + minSecond),
                    segmentCount: segments.length,
                    _segments: segments.map(s => ({
                        joinTime: s.joinTime,
                        leaveTime: s.leaveTime
                    })),
                    _meetingIndex: i,
                    _firstHalfEnd: firstEnd,
                    _secondHalfStart: secondStart
                });
            }
        }

        return records;
    },

    // ------------------------------------------------------------------

    _groupByParticipant(segments) {
        const groups = new Map();
        for (const seg of segments) {
            const key = seg.email || seg.fullName;
            if (!groups.has(key)) groups.set(key, []);
            groups.get(key).push(seg);
        }
        return groups;
    },

    _overlap(segStart, segEnd, rangeStart, rangeEnd) {
        const start = Math.max(segStart.getTime(), rangeStart.getTime());
        const end = Math.min(segEnd.getTime(), rangeEnd.getTime());
        return Math.max(0, (end - start) / 60000);
    },

    _round1(n) {
        return Math.round(n * 10) / 10;
    }
};

File: backend/attendance_normalization/aggregator.py
"""Meeting aggregation logic migrated from ``attendance/adapter/js/aggregator.js``.

This module intentionally keeps the "marker" problem outside of itself:
- the caller provides ``effective_start`` (blue marker)
- the caller provides ``break_point`` (yellow marker)

That keeps aggregation deterministic and easy to test.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Iterable


MAX_RECONNECT_GAP_MINUTES = 5.0


@dataclass(frozen=True)
class ZoomSegment:
    first_name: str
    last_name: str
    email: str
    full_name: str
    join_time: datetime
    leave_time: datetime


@dataclass(frozen=True)
class ZoomMeeting:
    course: str
    meeting_id: str
    start_time: datetime
    end_time: datetime
    duration_minutes: float
    segments: list[ZoomSegment]


@dataclass(frozen=True)
class AttendanceAggregationRecord:
    course: str
    meeting_id: str
    date: datetime
    meeting_start: datetime
    meeting_end: datetime
    meeting_duration: float
    first_name: str
    last_name: str
    email: str
    full_name: str
    minutes_first_half: float
    minutes_second_half: float
    duration_first_half: float
    duration_second_half: float
    total_minutes: float
    segment_count: int
    segments: list[tuple[datetime, datetime]]
    first_half_end: datetime
    second_half_start: datetime


def aggregate_meeting(
    meeting: ZoomMeeting,
    effective_start: datetime,
    break_point: datetime,
    effective_end: datetime | None = None,
) -> list[AttendanceAggregationRecord]:
    """Aggregate a single meeting into one record per participant.

    This mirrors the current JavaScript adapter behavior when explicit markers
    are already available.
    """

    if not meeting.segments:
        return []

    effective_meeting_end = effective_end or meeting.end_time
    first_half_end = break_point
    second_half_start = break_point

    duration_first_half = _minutes_between(effective_start, first_half_end)
    duration_second_half = _minutes_between(second_half_start, effective_meeting_end)

    records: list[AttendanceAggregationRecord] = []
    for participant_segments in _group_by_participant(meeting.segments).values():
        first = participant_segments[0]
        raw_intervals = [
            (segment.join_time, segment.leave_time)
            for segment in participant_segments
        ]
        attendance_intervals = _merge_nearby_intervals(raw_intervals)

        minutes_first_half = 0.0
        minutes_second_half = 0.0

        for segment_start, segment_end in attendance_intervals:
            minutes_first_half += _overlap_minutes(
                segment_start,
                segment_end,
                effective_start,
                first_half_end,
            )
            minutes_second_half += _overlap_minutes(
                segment_start,
                segment_end,
                second_half_start,
                effective_meeting_end,
            )

        records.append(
            AttendanceAggregationRecord(
                course=meeting.course,
                meeting_id=meeting.meeting_id,
                date=meeting.start_time,
                meeting_start=meeting.start_time,
                meeting_end=effective_meeting_end,
                meeting_duration=meeting.duration_minutes,
                first_name=first.first_name,
                last_name=first.last_name,
                email=first.email,
                full_name=first.full_name,
                minutes_first_half=_round1(minutes_first_half),
                minutes_second_half=_round1(minutes_second_half),
                duration_first_half=_round1(duration_first_half),
                duration_second_half=_round1(duration_second_half),
                total_minutes=_round1(minutes_first_half + minutes_second_half),
                segment_count=len(participant_segments),
                segments=raw_intervals,
                first_half_end=first_half_end,
                second_half_start=second_half_start,
            )
        )

    return records


def _group_by_participant(
    segments: Iterable[ZoomSegment],
) -> dict[str, list[ZoomSegment]]:
    groups: dict[str, list[ZoomSegment]] = {}
    for segment in segments:
        key = segment.email or segment.full_name
        groups.setdefault(key, []).append(segment)
    return groups


def _overlap_minutes(
    segment_start: datetime,
    segment_end: datetime,
    range_start: datetime,
    range_end: datetime,
) -> float:
    start = max(segment_start, range_start)
    end = min(segment_end, range_end)
    return max(0.0, (end - start).total_seconds() / 60)


def _merge_nearby_intervals(
    intervals: list[tuple[datetime, datetime]],
    max_gap: timedelta = timedelta(minutes=MAX_RECONNECT_GAP_MINUTES),
) -> list[tuple[datetime, datetime]]:
    if not intervals:
        return []

    ordered = sorted(intervals, key=lambda interval: interval[0])
    merged: list[list[datetime]] = [[ordered[0][0], ordered[0][1]]]

    for start, end in ordered[1:]:
        last = merged[-1]
        if start <= last[1] + max_gap:
            if end > last[1]:
                last[1] = end
        else:
            merged.append([start, end])

    return [(start, end) for start, end in merged]


def _minutes_between(start: datetime, end: datetime) -> float:
    return (end - start).total_seconds() / 60


def _round1(value: float) -> float:
    return round(value, 1)
